### PR TITLE
Do not parse `{displayName}` placeholder for `@ParameterizedTest` using `MessageFormat`

### PR DESCRIPTION
**Pull request by alemendoza-v**
## Overview

The `@ParameterizedTest` annotation was breaking name generation by not allowing `'` in the method names for Kotlin-based tests or in custom display names in geneeral. It also made it so that placeholders were not being replaced correctly.

To solve it, instead of replacing the `DISPLAY_NAME_PLACEHOLDER` before the `MessageFormat` was being evaluated, the `DISPLAY_NAME_PLACEHOLDER` is now replaced with another temporary placeholder, which is then replaced after the `MessageFormat` is evaluated.

Solves: #3235 

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Change is documented in the [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestNameFormatter.java [hunk 1]
```diff
@@ -72,7 +73,7 @@ private Object[] extractNamedArguments(Object[] arguments) {
 
 	private String prepareMessageFormatPattern(int invocationIndex, Object[] arguments) {
 		String result = pattern//
-				.replace(DISPLAY_NAME_PLACEHOLDER, this.displayName)//
+				.replace(DISPLAY_NAME_PLACEHOLDER, "__DISPLAY_NAME__")//
```
Comment (sbrannen) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestNameFormatter.java:76]: Please extract this to a constant.

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+class ParameterizedTestNameFormatterIntegrationTests {
+    @BeforeEach
+    fun setUp(info: TestInfo) {
+        System.out.printf(
+            "Class: %s, method: %s%nDisplay Name: %s%n%n",
+            info.testClass.orElseThrow(),
+            info.testMethod.orElseThrow(),
+            info.displayName
+        )
+    }
```
Comment (sbrannen) [inline junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt:28]: We do not print to SYSOUT in tests for debugging purposes.

So, please remove this method.

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    @MethodSource("methodSource")
+    @ParameterizedTest
+    fun `implicit'Name`(param: String) {
+        Assertions.assertNotNull(param)
```
Comment (sbrannen) [inline junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt:33]: Please static imports for `Assertions`.

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @MethodSource("methodSource")
+    @ParameterizedTest(name = "{0}")
+    fun `zero'Only`(param: String) {
+        Assertions.assertNotNull(param)
```
Comment (sbrannen) [inline junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt:39]: Please modify all test methods to accept `TestInfo` as an argument and then replace the `Assertions.assertNotNull(param)` assertions with something that asserts the actual expected display name.

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for submitting the PR.

I've requested a few minor changes.

**Comment by alemendoza-v** [comment]
@sbrannen Sure I'll get to work on these changes. Should the Release Notes entry include the basic change? Something like "Bug Fix: `@ParameterizedTest` now accepts single apostrophes inside the method names in Kotlin tests"?

**Comment by alemendoza-v** [comment]
@sbrannen Also,  should the new constant be inside the `Interface` or inside the `ParameterizedTest` class?

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt (hunk 3 above)
Comment (alemendoza-v) [reply junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt:33]: I think static imports in Kotlin do not exist :(

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestNameFormatter.java [hunk 5]
```diff
@@ -36,6 +36,7 @@ class ParameterizedTestNameFormatter {
 	private final String displayName;
 	private final ParameterizedTestMethodContext methodContext;
 	private final int argumentMaxLength;
+	private final String DISPLAY_NAME_TEMPORARY_PLACEHOLDER = "__DISPLAY_NAME__";
```
Comment (sbrannen) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/ParameterizedTestNameFormatter.java:39]: Please make this `static` and declare it under `ELLIPSIS`.

File: junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestNameFormatterTests.java [hunk 6]
```diff
@@ -60,6 +60,14 @@ void formatsDisplayName() {
...
 	}
 
+	@Test
+	void formatsDisplayNameWithApostrophe() {
+		var formatter = formatter(DISPLAY_NAME_PLACEHOLDER, "display'Zero");
+
+		assertEquals("display'Zero", formatter.format(1));
+		assertEquals("display'Zero", formatter.format(2));
+	}
+
```
Comment (sbrannen) [inline junit-jupiter-params/src/test/java/org/junit/jupiter/params/ParameterizedTestNameFormatterTests.java:70]:
```suggestion
	@Test
	void formatsDisplayNameWithApostrophe() {
		String displayName = "display'Zero";
		var formatter = formatter(DISPLAY_NAME_PLACEHOLDER, displayName);

		assertEquals(displayName, formatter.format(1));
		assertEquals(displayName, formatter.format(2));
	}

```

Please extract the `displayName` variable as above.

And please introduce the following test which demonstrates that the bug is not solely related to apostrophes.

```java
	@Test
	void formatsDisplayNameContainingFormatElements() {
		String displayName = "{enigma} {0} '{1}'";
		var formatter = formatter(DISPLAY_NAME_PLACEHOLDER, displayName);

		assertEquals(displayName, formatter.format(1));
		assertEquals(displayName, formatter.format(2));
	}
```

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.MethodOrderer
+import org.junit.jupiter.api.TestInfo
+import org.junit.jupiter.api.TestMethodOrder
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.MethodSource
+
+@TestMethodOrder(MethodOrderer.MethodName::class)
```
Comment (sbrannen) [inline junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt:17]:
```suggestion
```

You can safely remove this, since it is not necessary for these tests.

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+class ParameterizedTestNameFormatterIntegrationTests {
+
+    @MethodSource("methodSource")
```
Comment (sbrannen) [inline junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt:20]: Why did you choose the more complicated setup for `@MethodSource` with a companion object and `@JvmStatic`?

I'd likely go with `@ValueSource(strings = ["foo", "bar"])` (assuming that's the correct syntax for Kotlin).

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for making the previously requested changes.

I've requested a few additional changes after the latest review.

**Comment by sbrannen** [comment]
FYI: changes to the Javadoc and User Guide are not needed for this change, so I've updated the _Deliverables_ accordingly.

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt (hunk 3 above)
Comment (sbrannen) [reply junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt]: They do, but they don't need the keyword `static` like in Java.

See `org.junit.jupiter.params.aggregator.ArgumentsAccessorKotlinTests` for examples.

**Comment by sbrannen** [comment]
> should the new constant be inside the `Interface` or inside the `ParameterizedTest` class?

It's good as a private constant in `ParameterizedTestNameFormatter` as you have it now, since it is an internal implementation detail local to exactly that one class.

**Comment by sbrannen** [comment]
> Should the Release Notes entry include the basic change? Something like "Bug Fix: `@ParameterizedTest` now accepts single apostrophes inside the method names in Kotlin tests"?

There's no need to prefix it with "Bug Fix:". There's a [dedicated section for bug fixes in JUnit Jupiter](https://github.com/junit-team/junit5/blob/main/documentation/src/docs/asciidoc/release-notes/release-notes-5.10.0-M1.adoc#bug-fixes-1).

Regarding the content, it should be more generic in nature, explaining that the `{displayName}` placeholder is no longer parsed using `MessageFormat`, which means that a `@DisplayName` or Kotlin test method name can contain apostrophes and `MessageFormat` _elements_ (such as `{data}`).

See what you come up with, and if we deem it necessary we will polish the wording when merging the PR.

**Comment by alemendoza-v** [comment]
@sbrannen Thanks for reviewing, I'll get to work on the changes and the release notes entry :)

File: junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt (hunk 8 above)
Comment (alemendoza-v) [reply junit-jupiter-params/src/test/kotlin/ParameterizedTestNameFormatterIntegrationTests.kt:20]: The example that the author showed used that setup. But you're right I'll change it to a simpler setup.

**Comment by sbrannen** [comment]
This has been merged into `main` in 7b2688c8cbe6349656ef94492dceff6537d6a87c and revised in 745714463175df7198b018c446ea7a31fb19cd65.

In addition, since this PR was finalized before the `5.9.3` release, I backported this to `5.9.x` for inclusion in the upcoming `5.9.3` release.

Thanks

**Comment by alemendoza-v** [comment]
Thank you!
